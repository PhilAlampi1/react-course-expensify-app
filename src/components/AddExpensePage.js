import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'

// AddExpensePage needs to be a class-based component to avoid inline function (onSubmit)
// If you didn't do this, onSubmit would need to get recalculated on every render and re-render
// See commented out code below for old way.
export class AddExpensePage extends React.Component {
    onSubmit=(expense) => {
        this.props.addExpense(expense) // new way
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
            <h1>Add Expense</h1>
            <ExpenseForm 
                onSubmit={this.onSubmit}
            />
        </div>
        )
    }
}

// OLD WAY - created an inline function situation on Submit when changed for enzyme testing
// const AddExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm 
//             onSubmit={(expense) => {
//                 {/* props.dispatch(addExpense(expense)) */} // old way, prior to test consideration, see below.
//                 props.onSubmit(expense) // new way
//                 props.history.push('/')
//             }}
//         />
//     </div>
// )

// Because addExpense() is an imported function above, it's going to be hard to test as a 
// "spy" in enzyme. Instead, define and pass in the "mapDispatchToProps" argument to connect as shown below:
const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)