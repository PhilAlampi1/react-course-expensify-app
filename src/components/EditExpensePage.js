import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

// NOTE - need to add this. in front of all props for this to work
export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense)
        // this.props.dispatch(editExpense(this.props.expense.id, expense)) // inline function (bad)
        this.props.history.push('/')
    }
    onRemove = () => {
        this.props.startRemoveExpense({
            id: this.props.expense.id
        })
        // this. props.dispatch(removeExpense({ id: this.props.expense.id })) // inline function (bad)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
                </div>
            </div>
        )
    }
}

// const EditExpensePage = (props) => {
//     return (
//         <div>
//             <ExpenseForm
//                 expense={props.expense}
//                 onSubmit={(expense) => {
//                     props.dispatch(editExpense(props.expense.id, expense))
//                     props.history.push('/')
//                 }}
//             />
//             <button onClick={() => {
//                 props.dispatch(removeExpense({ id: props.expense.id }))
//                 props.history.push('/')
//             }}>Remove</button>
//         </div>
//     )
// }

// props.match.params.id is the URL params passed in
const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

// NOTE - below you pass in the arguement reference, not the value (e.g. don't try to call this.props stuff). Take care of the values
// when calling these methods above.
// NOTE - added the props argument because method definitions above required it (I think that's why?)
const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)