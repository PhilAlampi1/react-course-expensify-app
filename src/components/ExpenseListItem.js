import React from 'react'
import { Link } from 'react-router-dom'

// NOTE: you are passing an object into props.dispatch below
export const ExpenseListItem = (props) => (
    <div>
        <Link to={`/edit/${props.expense.id}`}>
            <h3>{props.expense.description}</h3>
        </Link>
        <p>{props.expense.amount} - {props.expense.createdAt}</p>
    </div>
)

// NO need to provide first argument set of connect since you are not drawing new data from state,
// you are using the prop.expense.id passed to the component above
export default ExpenseListItem