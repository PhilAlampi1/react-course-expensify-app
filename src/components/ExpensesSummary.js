import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import numeral from 'numeral'
import getExpensesTotal from '../selectors/expenses.total'
import getVisibleExpenses from '../selectors/expenses'

export const ExpensesSummary = (props) => (
    <div>
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    Viewing <span>{props.expensesCount}</span> expense{(props.expensesCount > 1 || props.expensesCount === 0) && 's'} totalling <span>{props.expensesTotal}</span>
                </h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    </div>
)

const mapStateToProps = (state) => ({
    expensesCount: getVisibleExpenses(state.expenses, state.filters).length,
    expensesTotal: numeral((getExpensesTotal(getVisibleExpenses(state.expenses, state.filters))) / 100).format('$0,0.00')
})

export default connect(mapStateToProps)(ExpensesSummary)