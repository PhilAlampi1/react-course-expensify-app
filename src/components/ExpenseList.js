import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses'

// Unconnected component
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                    props.expenses.map((e) => (
                        <ExpenseListItem key={e.id} expense={e} />
                    ))
                )
        }

    </div>
)

// Map state to props
const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}

// Put them together using "connect"
// Connected components are automatically reactive
export default connect(mapStateToProps)(ExpenseList)
