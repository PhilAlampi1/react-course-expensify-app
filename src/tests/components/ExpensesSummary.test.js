import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'

test('should render ExpensesSummary with no values passed', () => {
    const wrapper = shallow(<ExpensesSummary />)
    expect(wrapper).toMatchSnapshot()
})

test('should render ExpensesSummary with one item', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={9500} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render ExpensesSummary with multiple items', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={2} expensesTotal={9500} />)
    expect(wrapper).toMatchSnapshot()
})

// test('', () => {
    
// })