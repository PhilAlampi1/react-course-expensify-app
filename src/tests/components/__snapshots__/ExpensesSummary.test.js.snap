// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with multiple items 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Viewing 
        <span>
          2
        </span>
         expense
        s
         totalling 
        <span>
          9500
        </span>
      </h1>
      <div
        className="page-header__actions"
      >
        <Link
          className="button"
          replace={false}
          to="/create"
        >
          Add Expense
        </Link>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with no values passed 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Viewing 
        <span />
         expense
         totalling 
        <span />
      </h1>
      <div
        className="page-header__actions"
      >
        <Link
          className="button"
          replace={false}
          to="/create"
        >
          Add Expense
        </Link>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with one item 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Viewing 
        <span>
          1
        </span>
         expense
         totalling 
        <span>
          9500
        </span>
      </h1>
      <div
        className="page-header__actions"
      >
        <Link
          className="button"
          replace={false}
          to="/create"
        >
          Add Expense
        </Link>
      </div>
    </div>
  </div>
</div>
`;
